Like operator (#160)
Materials-Consortia/OPTIMADE: optimade.md

OPTIONAL features: 

The following comparison operators are OPTIONAL:

> * `identfier LIKE x`

Reviewer: ```suggestion
* `identifier LIKE x`

```

> * `identfier LIKE x`
> 
> * `identfier UNLIKE x`

Reviewer: ```suggestion
* `identifier LIKE pattern`: Is true if the property matches the provided `pattern`.

* `identifier UNLIKE x`: Is true if the property does not match the provided `pattern`.

```

Adds a bit of explanation matching the format for `CONTAINS` etc. above. I think `pattern` is clearer than `x` here, as `x` is used as a placeholder for the substrings above. Also fixes a typo in "`identifier`" - will also add a separate suggestion for this in isolation.

> * `identfier UNLIKE x`

Reviewer: ```suggestion
* `identifier UNLIKE x`

```

> * `identfier UNLIKE x`

Reviewer: Is there a strong preference for `UNLIKE` rather than just `NOT(identifier LIKE pattern)`? Is `UNLIKE` is an SQL construct that I have not come across? 

I guess the precedent for this is our inclusion of `KNOWN` and `UNKNOWN` -- I just don't think I like having two ways of doing different optional things, e.g., an implementation could choose to implement only `LIKE` and therefore has to implement `~LIKE` but can still choose to ignore `UNLIKE`. At least `KNOWN`/`UNKNOWN` are both mandatory. I guess we could add to the spec that if `LIKE` is implemented then `UNLIKE` must also be implemented.

Author: > Is there a strong preference for `UNLIKE` rather than just `NOT(identifier LIKE pattern)`? Is `UNLIKE` is an SQL construct that I have not come across?

UNLIKE is shorter, which is significant for URL-embedded querries...
 
> I guess the precedent for this is our inclusion of `KNOWN` and `UNKNOWN` -- I just don't think I like having two ways of doing different optional things, e.g., an implementation could choose to implement only `LIKE` and therefore has to implement `~LIKE` but can still choose to ignore `UNLIKE`. At least `KNOWN`/`UNKNOWN` are both mandatory. I guess we could add to the spec that if `LIKE` is implemented then `UNLIKE` must also be implemented.

I would say that if LIKE is implemented, UNLIKE also MUST be implemented; i.e. `x UNLIKE "%s%"` and `NOT(x LIKE "%s%"` MUST be synonyms – can we decide on that?  If the `LIKE` is implemented, then implementing `UNLIKE` does not require any extra effort.

* Support for x to be an identifier, rather than a string is OPTIONAL.

> If implemented, the "LIKE" operator MUST behave as the correspoding standard SQL operator. In particular,
> The `x` string MUST be interpreted as a pattern where an underscore character ('_', ASCII DEC 95, HEX 5F)
> matches any single character and a percent character ('%', ASCII DEC 37, HEX 25) matches an arbitrary
> sequence of characters (including zero characters).

Reviewer: ```suggestion
If implemented, the `LIKE` operator MUST behave as the corresponding standard SQL operator.
The `x` string MUST be interpreted as a string-matching pattern where an underscore character ('_', ASCII DEC 95, HEX 5F) matches any single character and a percent character ('%', ASCII DEC 37, HEX 25) matches an arbitrary sequence of characters (including zero characters).

```

Fix a few typos and enforces one line per sentence

Author: > @sauliusg makes a good point about the escapes. The text in the json schema standard points to the syntax standardized in [ECMA-262, section 21.2.1](https://262.ecma-international.org/11.0/#sec-regexp-regular-expression-objects) which says the syntax is "modeled" after Perl 5. However, the json schema standard then says - on SHOULD level - that the expressions should be "limit[ed] [...] to the following regular expression tokens", which doesn't actually include any description of the escape token.
> 
> This is likely an oversight. In the ECMA standard it is handled in the grammar where an `Atom` can be `\` plus an `AtomEscape`, which is then further discussed in [21.2.2.8.1](https://262.ecma-international.org/11.0/#sec-regexp-regular-expression-objects).

If we describe, as suggested, the RE syntax in the OPTIMADE independently, we can strengthen the requirements as we see it fit. This is a good opportunity to correct the oversight, if it really was.

> If operator "UNLIKE" is supported, the bahavior of this oprtator MUST be the negation of the "LIKE" operator; i.e.

Reviewer 2: ```suggestion
If operator "UNLIKE" is supported, the behavior of this operator MUST be the negation of the "LIKE" operator; i.e.

```

Reviewer: ```suggestion
If the `UNLIKE` operator is supported, the behavior of this operator MUST be the negation of the "LIKE" operator; i.e.

```

based on the comment above, one more reformatting/wording change

> an expression `(property UNLIKE "value")" must behave exactly as `(NOT (property LIKE "value"))`.

Reviewer 2: ```suggestion
expression `(property UNLIKE "value")` must behave exactly as `(NOT (property LIKE "value"))`.

```

OPTIONAL features: 

The following comparison operators are OPTIONAL:

* `identfier LIKE x`

* `identfier UNLIKE x`

* Support for x to be an identifier, rather than a string is OPTIONAL.

If implemented, the "LIKE" operator MUST behave as the correspoding standard SQL operator. In particular,
The `x` string MUST be interpreted as a pattern where an underscore character ('_', ASCII DEC 95, HEX 5F)
matches any single character and a percent character ('%', ASCII DEC 37, HEX 25) matches an arbitrary

> sequence of characters (including zero characters).

Reviewer 3: Shouldn't then we also explain how to escape these characters, and how to escape the escape character? (should be `\` IIRC

Reviewer 4: Good catch. [MySQL specification](https://dev.mysql.com/doc/refman/5.5/en/string-comparison-functions.html) says that to search for `\` in `LIKE` one would have to write `\\\\` (instead of `\\`, which is common way to escape the escaping sequences in C and the like). Which will we choose?

Author: Thanks @giovannipizzi & @merkys for spotting this deficiency; I'll look up the existing escape rules and suggest a version in the text.

I agree with @giovannipizzi that we can use single backslash to escape characters, but we need to think over all possible combinations. I'll propose a variant in a while.
